src/clj/game/core/costs.clj

```diff
@@ -266,8 +266,22 @@
 
 (defn pay-credits
   [state side eid card amount]
-  (swap! state update-in [:stats side :spent :credit] (fnil + 0) amount)
-  (complete-with-result state side eid (deduct state side [:credit amount])))
+  (case (:source-type eid)
+    :runner-install
+    (letfn [(provider-func []
+              (filter #((or (-> % card-def :interactions :pay-credits) (req false)) state side eid % [card]) (all-active-installed state :runner)))]
```

Comment:
This feels inefficient. Maybe generate the list of `all-installed` once, and then `(filter #(when-let [pay-credits (-> % card-def :interactions :pay-credits)] (pay-credits state side eid % [card])) the-installed-cards)` The game-state shouldn't be changing when you're paying for something, so you only need to generate the list of available cards once, and then making the anonymous function use `when-let` means we won't even execute the full function for any card that lacks `:pay-credits`.

Reply:
Wait a minute, actually, you should only call this whole thing once anyways! Since you've moved the `(< 0 (:recurring card))` check into the prompt, you can remove it from the card def. If we need to check things other than `:recurring`, we could pass that in as an argument to the select prompt, but as it is right now, you're checking each card twice to see if it has enough `:recurring` credits.

Reply:
We need to check providers every time `pick-credit-providing-cards` is called, since cards like Net Mercur exist. Imagine you're using a credit off of Cloak. Now Net Mercur got a credit, but wasn't on the initial list of providers.
I was calling it more often though and have changed that.

Reply:
Thinking more about it, the problem started with me comparing whole cards instead of `:cid`. Fixed that as well.

Reply:
Net Mercur doesn’t proc until the whole cost is paid, so you can’t spend on cloak and then also spend off Net Mercur for the same cost.